dtls.c

```diff
@@ -1608,7 +1608,7 @@ dtls_0_send_hello_verify_request(dtls_context_t *ctx,
 			     dtls_ephemeral_peer_t *ephemeral_peer,
 			     uint8 *data, size_t data_length)
 {
-  uint8 buf[DTLS_RH_LENGTH + DTLS_HS_LENGTH + data_length];
+  uint8 buf[DTLS_RH_LENGTH + DTLS_HS_LENGTH + sizeof(uint16) + sizeof(uint8) + DTLS_COOKIE_LENGTH];
```

Comment:
Could you give some explanation about this change?

Why do you consider to replace the `data_length` by `sizeof(uint16) + sizeof(uint8) + DTLS_COOKIE_LENGTH`? Does your compiler requires constant values here?

Reply:
> Does your compiler requires constant values here?

Yes, exactly, that was the reason for the change. The value for `data_length` is taken from here, therefore it still works:

https://github.com/eclipse/tinydtls/blob/bda40789a7c280f248eeca6d09ddd624cdaf5dc8/dtls.c#L1977-L1985

The current solution is, of course, very provisional. Would it make sense to turn `data_length` into a constant? Or do you see a better alternative?

Reply:
@obgm 

FMPOV, if that is required for some compilers (maybe there are also compile switches to enable the feature?) then it seems to be reasonable. If we go for that, I think we should remove the obsolete parameter.

Reply:
Yes, in my experience, some compilers cannot handle dynamic stack allocation.

Reply:
I check the code (and found it's mainly my changes for the stateless hello).

My proposal:

Replace `sizeof(uint16) + sizeof(uint8) + DTLS_COOKIE_LENGTH` by `DTLS_HV_LENGTH + DTLS_COOKIE_LENGTH`. Leave the parameter, because it contains the number of bytes to be copied from buf. Maybe an assert helps then to align the sizes of the copied bytes and `buf`.

(Before that stateless change, `dtls_send_multi` was used, which allocated the `sendbuf[DTLS_MAX_BUF]`. Maybe, in order to support `DTLS_CONSTRAINED_STACK`, we switch to that size/approach?)

Reply:
> Replace sizeof(uint16) + sizeof(uint8) + DTLS_COOKIE_LENGTH by DTLS_HV_LENGTH + DTLS_COOKIE_LENGTH. Leave the parameter, because it contains the number of bytes to be copied from buf. Maybe an assert helps then to align the sizes of the copied bytes and buf.

Thanks, that looks good :) I applied your suggestion in a99838f830b71ae4f38fa6a20f97bf8e033ed52e.

> (Before that stateless change, dtls_send_multi was used, which allocated the sendbuf[DTLS_MAX_BUF]. Maybe, in order to support DTLS_CONSTRAINED_STACK, we switch to that size/approach?)

Do you mean that we should use the following instead, changing `buf` to `sendbuf`?

```c
#ifndef DTLS_CONSTRAINED_STACK
  unsigned char sendbuf[DTLS_MAX_BUF];
#endif /* ! DTLS_CONSTRAINED_STACK */
```

Reply:
> Do you mean that we should use the following instead, changing buf to sendbuf?

Not with this PR. 

I also consider, to "allocate" the `buf` in `dtls_0_verify_peer`, I will check that later.